pkg/logging/logger.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logging
+
+import (
+	"context"
+
+	"github.com/pulumi/pulumi/pkg/diag"
+	"github.com/pulumi/pulumi/pkg/resource"
+	"github.com/pulumi/pulumi/pkg/resource/provider"
+)
+
+// Logger wraps a time-ordered log set to allow batched logging of unique messages.
+type Logger struct {
+	messages *TimeOrderedLogSet
+	index    int
+}
+
+// NewLogger returns an initialized Logger.
+func NewLogger() *Logger {
```

Comment:
I forgot the outcome of the last time we discussed this, so I apologize, but: we explicitly chose not to implement this in pulumi/pulumi, right? Why was that again? (I think this was my suggestion.)

Reply:
I don't recall that discussion, but it does seem like this could be useful in the core engine.